src/data/equiv/derangements/finite.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2021 Henry Swanson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henry Swanson
+-/
+import data.equiv.basic
+import data.equiv.derangements.basic
+import data.fintype.basic
+import data.fintype.card
+import data.set.finite
+
+import tactic.ring
+import tactic.zify
+
+/-!
+# Derangements on fintypes
+
+This file contains lemmas that describe the cardinality of `derangements α` when `α` is a fintype.
+
+# Main definitions
+
+  - `num_derangements n`: The number of derangements on an n-element set. For concreteness, `fin n`
+    is used.
+
+  - `num_derangements_invariant`: A lemma stating that the number of derangements on a type `α`
+    depends only on the cardinality of `α`.
+
+  - `num_derangements_recursive`: A lemma giving a recursive expression for `num_derangements (n+2)`
+    in terms of `num_derangements n` and `num_derangements (n+1)`.
+
+  - `num_derangements_sum`: A lemma giving an expression for `num_derangements n` in terms of
+    factorials.
+-/
+
+open equiv fintype perm
+open_locale big_operators
+
+variables {α : Type*} [decidable_eq α] [fintype α]
+
+instance : decidable_pred (derangements α) := λ _, fintype.decidable_forall_fintype
+
+instance : fintype (derangements α) :=
+begin
+  rw [derangements],
+  apply_instance
+end
+
+/-- The number of derangements on an `n`-element set. This definition is bad for
+    computation though, use `num_derangements` instead. -/
+def num_derangements_aux (n : ℕ) : ℕ := card (derangements (fin n))
```

Comment:
I'd be inclined to remove this definition entirely and inline it into the few places where it is used

Reply:
I've gotten rid of this now, and you're right, it's nicer.